[lldb][DWARFASTParserClang] Added a check for the specialization existence

tgs-sc opened the pull request:
[lldb][DWARFASTParserClang] Added a check for the specialization existence

    While debugging an application with incorrect dwarf information, where
    DW_TAG_template_value_parameter was lost, I found that lldb does not
    check that the corresponding specialization exists. As a result, at the
    stage when ASTImporter works, the type is completed in such a way that
    it inherits from itself. And during the calculation of layout, an
    infinite recursion occurs. To catch this error, I added a corresponding check
    at the stage of restoring the type from dwarf information. I also added a
    trivial assert in clang to check that the class does not inherit from itself.

tgs-sc commented:
@Michael137, is it OK?

tgs-sc commented:
@JDevlieghere, can you also please look at this?

tgs-sc commented:
@Michael137, can you please take a look as a week has passed? I added unittest and updated godbolt reproduction.

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -1873,6 +1874,24 @@ DWARFASTParserClang::ParseStructureLikeDIE(const SymbolContext &sc,
+      class_template_decl->AddSpecialization(class_specialization_decl,
+                                             InsertPos);
+    else {
+      module_sp->ReportError("SymbolFileDWARF({0:p}) - Specialization for "
```
I don't think we should be reporting this to the user console. Particularly, a user wouldn't really be aware of what a `Specialization for clang::ClassTemplateDecl` is and what the issue is with it already existing.

Lets just `LogConsole` here for now

Michael137 commented on lldb/unittests/SymbolFile/DWARF/DWARFASTParserClangTests.cpp:
```diff
@@ -599,6 +599,39 @@ TEST_F(DWARFASTParserClangTests, TestDefaultTemplateParamParsing) {
+
+  llvm::SmallVector<lldb::TypeSP, 3> types;
+  for (DWARFDIE die : cu_die.children()) {
+    if (die.Tag() == DW_TAG_structure_type) {
```
How about also checking that:
```
llvm::StringRef(die.GetName()).starts_with("_Optional_payload")
```
So you don't parse the `struct Type`. Then you just assert that one of the types is `nullptr` and the other isn't. Makes the test a bit clearer

tgs-sc replied:
Addressed

Michael137 commented on lldb/unittests/SymbolFile/DWARF/Inputs/DW_AT_spec_decl_exists-test.yaml:
```diff
@@ -0,0 +1,675 @@
+# }
+#
+# YAML generated on Linux using obj2yaml on the above program
+# compiled with G++.
```
Lets add a comment saying this is malformed DWARF that is missing `DW_TAG_template_value_parameter` entries, which is important for the test because that makes the two specializations structurally match in the AST.

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -1873,6 +1874,24 @@ DWARFASTParserClang::ParseStructureLikeDIE(const SymbolContext &sc,
+        !class_template_decl->findSpecialization(args, InsertPos))
+      // Add this specialization to the class template.
+      class_template_decl->AddSpecialization(class_specialization_decl,
+                                             InsertPos);
```
Not sure I understand. What's preventing us from calling `findSpecialization` inside `CreateClassTemplateSpecializationDecl `?

tgs-sc committed:
[lldb][DWARFASTParserClang] Added a check for the specialization existence

While debugging an application with incorrect dwarf information, where
DW_TAG_template_value_parameter was lost, I found that lldb does not
check that the corresponding specialization exists. As a result, at the
stage when ASTImporter works, the type is completed in such a way that
it inherits from itself. And during the calculation of layout, an
infinite recursion occurs. To catch this error, I added a corresponding check
at the stage of restoring the type from dwarf information.

tgs-sc commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -1873,6 +1873,19 @@ DWARFASTParserClang::ParseStructureLikeDIE(const SymbolContext &sc,
     clang_type =
         m_ast.CreateClassTemplateSpecializationType(class_specialization_decl);
 
+    // If we got nullptr, then such specialization already exists.
```
Addressed

Michael137 reviewed (approved):
LGTM (module remaining comments)

tgs-sc commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -1873,6 +1873,18 @@ DWARFASTParserClang::ParseStructureLikeDIE(const SymbolContext &sc,
+            "SymbolFileDWARF({0:p}) - Specialization for "
+            "clang::ClassTemplateDecl({1:p}) already exists.",
+            static_cast<void *>(this),
+            static_cast<void *>(class_template_decl));
```
Addressed

tgs-sc commented on lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:
```diff
@@ -1676,6 +1676,14 @@ TypeSystemClang::CreateClassTemplateSpecializationDecl(
+                                           InsertPos);
+  } else
+    // Specialization exists, so return nullptr.
+    return nullptr;
```
Addressed

tgs-sc commented:
@Michael137, do you think that this PR(https://github.com/llvm/llvm-project/pull/154134) is still needed?

Michael137 commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -1873,6 +1873,17 @@ DWARFASTParserClang::ParseStructureLikeDIE(const SymbolContext &sc,
+      if (log) {
+        dwarf->GetObjectFile()->GetModule()->LogMessage(
+            log,
+            "SymbolFileDWARF({0:p}) - Specialization for "
+            "clang::ClassTemplateDecl({1}, {2:p}) already exists.",
```
```suggestion
            "SymbolFileDWARF({0:p}) - Failed to create specialization for "
            "clang::ClassTemplateDecl({1}, {2:p}).",
```

Michael137 commented on lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:
```diff
@@ -1676,6 +1676,13 @@ TypeSystemClang::CreateClassTemplateSpecializationDecl(
+  void *InsertPos = nullptr;
+  if (class_template_decl->findSpecialization(args, InsertPos))
+    return nullptr;
+  // Add this specialization to the class template.
```
```suggestion
```

tgs-sc commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:
```diff
@@ -1873,6 +1873,17 @@ DWARFASTParserClang::ParseStructureLikeDIE(const SymbolContext &sc,
     clang_type =
         m_ast.CreateClassTemplateSpecializationType(class_specialization_decl);
```
Addressed

tgs-sc commented:
@Michael137, can you please check what I suggested as a fix?

tgs-sc commented:
@Michael137, can you please rerun CI in this PR? I guess my fix is gonna work.

tgs-sc commented on lldb/source/Plugins/TypeSystem/Clang/TypeSystemClang.cpp:
```diff
@@ -959,6 +959,9 @@ CompilerType TypeSystemClang::GetBuiltinTypeForDWARFEncodingAndBitSize(
     if (type_name == "long double" &&
         QualTypeMatchesBitSize(bit_size, ast, ast.LongDoubleTy))
       return GetType(ast.LongDoubleTy);
+    if (type_name == "__bf16" &&
```
> This needs to be a separate PR

OK, I will create one. With a test?

Michael137 replied:
Ideally yes

tgs-sc commented:
I close it as it is actually merged (42a5c3d74a1a19)

tgs-sc commented:
Oh, I am sorry, but it is not merged as I checked. @Michael137, can you merge it please?

DavidSpickett commented:
I see an approval but also a sea of outstanding comments later on. @Michael137 please merge this if you know it's fine as is, or highlight anything that remains to be done.

(they are likely at the event in the US, so expect some delay)

Michael137 reviewed (changes requested):
Once last question is resolved I'll go ahead and merge
